app/internal/features.py

```diff
@@ -0,0 +1,209 @@
+) -> Feature:
+    feature.name = feature_dict["name"]
+    feature.route = feature_dict["route"]
+    feature.description = feature_dict["description"]
+    feature.creator = feature_dict["creator"]
+    session.commit()
+    return feature
+
+
+def is_feature_enabled(
+    user: User,
+    feature: Feature,
+    session: SessionLocal = Depends(get_db),
+) -> bool:
+    enabled_features = get_user_enabled_features(session=session, user=user)
+    return any(ef.id == feature.id for ef in enabled_features)
+
+
+async def is_access_allowd(request: Request, route: str) -> bool:
+
+    session = SessionLocal()
+
+    # To get current user.
+    # Note: can't use dependency beacause its designed for routes only.
+    # Needed to it manualy.
+    jwt = await get_authorization_cookie(request=request)
+    user = await current_user_from_db(request=request, jwt=jwt, db=session)
+
+    feature = session.query(Feature).filter_by(route=route).first()
+
+    if feature is None:
+        # in case there is no feature exists in the database that match the
+        # route that gived by to the request.
+        return True
+
+    user_pref = (
+        session.query(UserFeature)
+        .filter_by(feature_id=feature.id, user_id=user.id)
+        .first()
+    )
+
+    return user_pref is not None and user_pref.is_enable
+
+
+def create_feature(
+    name: str,
+    route: str,
+    description: str,
+    creator: str = None,
+    db: SessionLocal = Depends(),
+) -> Feature:
+    """Creates a feature."""
+    db = SessionLocal()
+    return create_model(
+        db,
+        Feature,
+        name=name,
+        route=route,
+        creator=creator,
+        description=description,
+    )
+
+
+def create_user_feature_association(
+    db: SessionLocal,
+    feature_id: int,
+    user_id: int,
+    is_enable: bool,
+) -> UserFeature:
+    """Creates an association."""
+    return create_model(
+        db,
+        UserFeature,
+        user_id=user_id,
+        feature_id=feature_id,
+        is_enable=is_enable,
+    )
+
+
+def get_user_enabled_features(
+    user_id: int,
+    session: SessionLocal = Depends(get_db),
+) -> List[Feature]:
+    enabled = []
+    user_prefs = session.query(UserFeature).filter_by(user_id=user_id).all()
+
+    for pref in user_prefs:
+        if pref.is_enable:
+            feature = (
+                session.query(Feature).filter_by(id=pref.feature_id).first()
+            )
+            enabled.append(feature)
+
+    return enabled
+
+
+async def get_user_uninstalled_features(
+    request: Request,
+    session: SessionLocal = Depends(get_db),
+) -> List[Feature]:
+    uninstalled = []
+    all_features = session.query(Feature).all()
+
+    # To get current user.
+    # Note: can't use dependency beacause its designed for routes only.
+    # Needed to it manualy.
+    jwt = await get_authorization_cookie(request=request)
+    user = await current_user_from_db(request=request, jwt=jwt, db=session)
```

Comment:
Why you need this? Can't you just use the JWT token? Retrieving the from the database is expansive

Reply:
I can, I talked with koby and he told me to use `current_user` instead of `current_user_from_db`.
`current_user` don't retrieving from the database :)